File: codes/leetcode/AddOneRowtoTree.cpp
/*
************************************************
username    :   smmehrab
fullname    :   s.m.mehrabul islam
email       :   mehrab.24csedu.001@gmail.com
institute   :   university of dhaka, bangladesh
session     :   2017-2018
************************************************
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int v, int d) {
        if (d == 1) {
            TreeNode* node = new TreeNode(v);
            node->left = root;
            return node;
        }
        
        TreeNode* node;
        TreeNode* left;
        TreeNode* right;
        queue<TreeNode*> q;
        q.push(root);
        
        int level = 1;
        while (level < d-1) {
            int levelNodesCount = q.size();
            while(levelNodesCount--) {
                node = q.front();
                q.pop();
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);                
            }
            level++;
        }
        
        while (!q.empty()) {
            node = q.front();
            q.pop();
            
            left = node->left;
            right = node->right;
            
            node->left = new TreeNode(v);
            node->left->left = left;
            
            node->right = new TreeNode(v);
            node->right->right = right;
        }
        
        return root;
    }
};

File: codes/leetcode/AddOneRowtoTree.py
# ************************************************
# username  :   smmehrab
# fullname  :   s.m.mehrabul islam
# email     :   smmehrabul-2017614964@cs.du.ac.bd
# institute :   university of dhaka, bangladesh
# reg       :   2017614964
# ************************************************

from collections import deque

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):

    def addOneRow(self, root, val, depth):
        """
        :type root: TreeNode
        :type val: int
        :type depth: int
        :rtype: TreeNode
        """

        if depth == 1:
            root = TreeNode(val, root, None)

        else:
            queue = deque()
            queue.append(root)
            current_depth = 1
            while queue and current_depth<depth-1:
                current_depth_nodes_count = len(queue)
                while current_depth_nodes_count:
                    current_depth_nodes_count -= 1
                    node = queue.popleft()
                    if node.left:
                        queue.append(node.left)
                    if node.right:
                        queue.append(node.right)
                current_depth += 1

            while queue:
                node = queue.popleft()
                node.left = TreeNode(val, node.left, None)
                node.right = TreeNode(val, None, node.right)

        return root